### PR TITLE
Dart constructor tearoff support.

### PR DESCRIPTION
**Pull request by jwren**
See ConstructorTearoffs.dart test, with the exception of the last expression, `var f2 = C<num>.new;` I have the Dart.bnf parser working with the new syntax.

@alexander-doroshko -- thanks for any insights to getting around the BNF parsing of this expression as a new expression or call expression instead of a compare expression.

### CODE REVIEW

File: Dart/src/com/jetbrains/lang/dart/Dart.bnf [hunk 1]
```diff
@@ -313,19 +313,19 @@ private functionNative ::= 'native' (stringLiteralExpression ';' | ';' | stringL
 
 methodDeclaration ::= metadata* ('external' | 'static' | 'const')* methodDeclarationPrivate initializers? (';' | functionBodyOrNative | redirection)?
 {pin=3 mixin="com.jetbrains.lang.dart.psi.impl.AbstractDartMethodDeclarationImpl" implements="com.jetbrains.lang.dart.psi.DartComponent"}
-private methodDeclarationPrivate ::= returnType <<methodNameWrapper>> typeParameters? formalParameterList | !untypedFunctionType <<methodNameWrapper>> typeParameters? formalParameterList // todo remove, use functionSignature as in spec
+private methodDeclarationPrivate ::= returnType (<<methodNameWrapper>> | 'new') typeParameters? formalParameterList | !untypedFunctionType (<<methodNameWrapper>> | 'new') typeParameters? formalParameterList // todo remove, use functionSignature as in spec
```
Comment (alexander-doroshko) [inline Dart/src/com/jetbrains/lang/dart/Dart.bnf:316]: It's suspicious that here's no '.' before 'new'. Are you sure this change is needed? Tests don't fail if I revert changes in this line.

**Comment by alexander-doroshko** [comment]
Thanks for the PR.
The remaining part of work is tricky indeed. I'm afraid I don't have an answer yet.

File: Dart/src/com/jetbrains/lang/dart/Dart.bnf (hunk 1 above)
Comment (jwren) [reply Dart/src/com/jetbrains/lang/dart/Dart.bnf]: You are correct, the syntax for named constructors kicks in making this change unnecessary, reverting here.

File: Dart/src/com/jetbrains/lang/dart/Dart.bnf [hunk 2]
```diff
@@ -313,19 +313,19 @@ private functionNative ::= 'native' (stringLiteralExpression ';' | ';' | stringL
 
 methodDeclaration ::= metadata* ('external' | 'static' | 'const')* methodDeclarationPrivate initializers? (';' | functionBodyOrNative | redirection)?
 {pin=3 mixin="com.jetbrains.lang.dart.psi.impl.AbstractDartMethodDeclarationImpl" implements="com.jetbrains.lang.dart.psi.DartComponent"}
-private methodDeclarationPrivate ::= returnType <<methodNameWrapper>> typeParameters? formalParameterList | !untypedFunctionType <<methodNameWrapper>> typeParameters? formalParameterList // todo remove, use functionSignature as in spec
+private methodDeclarationPrivate ::= returnType <<methodNameWrapper>> typeParameters? formalParameterList | !untypedFunctionType (<<methodNameWrapper>> | 'new') typeParameters? formalParameterList // todo remove, use functionSignature as in spec
```
Comment (alexander-doroshko) [inline Dart/src/com/jetbrains/lang/dart/Dart.bnf:316]: There were two `'new'` additions in this line =). I removed the second one as well before merging.

**Comment by alexander-doroshko** [comment]
Merged this wip. Tracking issue remains open: https://youtrack.jetbrains.com/issue/WEB-52293

**Pull request closed without merging**